### ISSUE TITLE
False positive when type is changed from interface to implementation

### ISSUE THREAD

**Issue by Wirone** [Roave/BackwardCompatibilityCheck#437]
ℹ️ Using `roave/backward-compatibility-check` 5.0

Today I've tested this library against our package that I've made changes in and I think it reports false-positive errors like:

```
[BC] CHANGED: Type documentation for property XF\Kernel\Controller\BaseController#$request changed from \Psr\Http\Message\ServerRequestInterface to \XF\Http\Request\ServerRequestProxy
[BC] CHANGED: The return type of XF\Kernel\Controller\BaseController#getRequest() changed from Psr\Http\Message\ServerRequestInterface to XF\Http\Request\ServerRequestProxy
```

In fact, `ServerRequestProxy` implements `ServerRequestInterface` and adds some more custom methods. So basically we're keeping 100% of previous API, only extending it.

**Comment by Ocramius** [comment]
For this one:

```
[BC] CHANGED: Type documentation for property XF\Kernel\Controller\BaseController#$request changed from \Psr\Http\Message\ServerRequestInterface to \XF\Http\Request\ServerRequestProxy
```

This type change is a BC break: writing a `ServerRequestInterface` to `BaseController#$request` will now lead to a crash.

```php
<?php

class ServerRequestInterface {}
class ServerRequestProxy extends ServerRequestInterface {}

class BaseController {
    public ServerRequestProxy $request;
}

$controller = new BaseController();

$controller->request = new ServerRequestProxy(); // ok
$controller->request = new ServerRequestInterface(); // crash
```
https://3v4l.org/p6CTt

```
[BC] CHANGED: The return type of XF\Kernel\Controller\BaseController#getRequest() changed from Psr\Http\Message\ServerRequestInterface to XF\Http\Request\ServerRequestProxy
```

This means that all class extending `BaseController` and declaring `BaseController#getRequest()` will need to be adjusted

```php
<?php

class ServerRequestInterface {}
class ServerRequestProxy extends ServerRequestInterface {}

class BaseController {
    public function getRequest(): ServerRequestProxy {}
}

class PreExistingController extends BaseController {
    public function getRequest(): ServerRequestInterface {}
}
```

https://3v4l.org/t5t26

**Comment by Ocramius** [comment]
Solutions:

 1. make `$request` property `private` (and generally, design `private` first)
 2. make `getRequest()` `final` (and design with final API as much as possible)

**Comment by Wirone** [comment]
`BaseController::$request` was initialised in `final` constructor but yeah, property itself was `protected`, so writing was allowed in child classes and it would be BC-break then. Thanks for pointing it out with useful explanation 🍻

**Comment by Ocramius** [comment]
I can totally endorse making everything `final`, `private` and `@internal`, when possible :P
